File: scripts/promote-founding-candidates.mjs
import fs from 'node:fs';
import path from 'node:path';

const root = process.cwd();
const reviewPath = path.join(root, 'research', 'revisit-lists', 'founding-promotion-review.csv');
const publicFoundingPath = path.join(root, 'public', 'company_founding_dates.csv');
const allowedRecommendation = 'PROMOTE_CANDIDATE';

const OUTPUT_HEADERS = ['company', 'founding_date', 'date_basis', 'source', 'confidence', 'notes'];

const csvEscape = (value) => {
  const text = String(value ?? '');
  return /[",\n]/.test(text) ? `"${text.replace(/"/g, '""')}"` : text;
};

const parseCsv = (text) => {
  const rows = [];
  let row = [];
  let cell = '';
  let inQuotes = false;

  for (let i = 0; i < text.length; i += 1) {
    const ch = text[i];
    const next = text[i + 1];
    if (ch === '"') {
      if (inQuotes && next === '"') {
        cell += '"';
        i += 1;
      } else {
        inQuotes = !inQuotes;
      }
    } else if (ch === ',' && !inQuotes) {
      row.push(cell);
      cell = '';
    } else if ((ch === '\n' || ch === '\r') && !inQuotes) {
      if (ch === '\r' && next === '\n') i += 1;
      row.push(cell);
      if (row.some((value) => value.trim())) rows.push(row);
      row = [];
      cell = '';
    } else {
      cell += ch;
    }
  }

  if (cell.length || row.length) {
    row.push(cell);
    if (row.some((value) => value.trim())) rows.push(row);
  }

  return rows;
};

const rowsFromCsv = (file) => {
  const rows = parseCsv(fs.readFileSync(file, 'utf8'));
  const headers = rows[0].map((header) => header.trim());
  return rows.slice(1).map((cols) => Object.fromEntries(headers.map((header, i) => [header, cols[i] ?? ''])));
};

const companyKey = (value) => String(value || '').trim().toLowerCase();
const normalizeCandidateDate = (value) => {
  const text = String(value || '').trim();
  let match = text.match(/^(\d{4})-00-00$/);
  if (match) return match[1];
  match = text.match(/^(\d{4})-(\d{2})-00$/);
  if (match) return match[2] === '00' ? match[1] : `${match[1]}-${match[2]}`;
  match = text.match(/^(\d{4})-00$/);
  if (match) return match[1];
  return text;
};

const existingRows = rowsFromCsv(publicFoundingPath);
const existingCompanies = new Set(existingRows.map((row) => companyKey(row.company)));
const reviewRows = rowsFromCsv(reviewPath);

const promotedRows = [];
for (const row of reviewRows) {
  if (row.recommendation !== allowedRecommendation) continue;
  if (!row.company || !row.founding_date || existingCompanies.has(companyKey(row.company))) continue;
  existingCompanies.add(companyKey(row.company));
  promotedRows.push({
    company: row.company,
    founding_date: normalizeCandidateDate(row.founding_date),
    date_basis: row.date_basis || 'operating_founding',
    source: row.source || '',
    confidence: row.confidence || '',
    notes: [row.notes, row.url ? `Source URL: ${row.url}` : ''].filter(Boolean).join('; '),
  });
}

const outputRows = [...existingRows, ...promotedRows];
fs.writeFileSync(
  publicFoundingPath,
  [OUTPUT_HEADERS, ...outputRows.map((row) => OUTPUT_HEADERS.map((header) => row[header] ?? ''))]
    .map((row) => row.map(csvEscape).join(','))
    .join('\n'),
);

console.log(JSON.stringify({
  promoted: promotedRows.length,
  total_live_rows: outputRows.length,
}, null, 2));

File: scripts/promote_founding_candidates.py
#!/usr/bin/env python3
"""Python port of scripts/promote-founding-candidates.mjs (no Node available in this env)."""
import csv
import json
import os
import re

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
REVIEW_PATH = os.path.join(ROOT, 'research', 'revisit-lists', 'founding-promotion-review.csv')
PUBLIC_FOUNDING_PATH = os.path.join(ROOT, 'public', 'company_founding_dates.csv')
ALLOWED_RECOMMENDATION = 'PROMOTE_CANDIDATE'
OUTPUT_HEADERS = ['company', 'founding_date', 'date_basis', 'source', 'confidence', 'notes']


def csv_escape(value):
    text = '' if value is None else str(value)
    if re.search(r'[",\n]', text):
        return '"' + text.replace('"', '""') + '"'
    return text


def company_key(value):
    return str(value or '').strip().lower()


def normalize_candidate_date(value):
    text = str(value or '').strip()
    m = re.match(r'^(\d{4})-00-00$', text)
    if m:
        return m[1]
    m = re.match(r'^(\d{4})-(\d{2})-00$', text)
    if m:
        return m[1] if m[2] == '00' else f'{m[1]}-{m[2]}'
    m = re.match(r'^(\d{4})-00$', text)
    if m:
        return m[1]
    return text


def rows_from_csv(path):
    with open(path, newline='', encoding='utf-8') as f:
        return list(csv.DictReader(f))


def main():
    existing_rows = rows_from_csv(PUBLIC_FOUNDING_PATH)
    existing_companies = {company_key(r['company']) for r in existing_rows}
    review_rows = rows_from_csv(REVIEW_PATH)

    promoted_rows = []
    for row in review_rows:
        if row.get('recommendation') != ALLOWED_RECOMMENDATION:
            continue
        if not row.get('company') or not row.get('founding_date') or company_key(row['company']) in existing_companies:
            continue
        existing_companies.add(company_key(row['company']))
        promoted_rows.append({
            'company': row['company'],
            'founding_date': normalize_candidate_date(row['founding_date']),
            'date_basis': row.get('date_basis') or 'operating_founding',
            'source': row.get('source') or '',
            'confidence': row.get('confidence') or '',
            'notes': '; '.join(filter(None, [row.get('notes', ''), f"Source URL: {row['url']}" if row.get('url') else ''])),
        })

    output_rows = existing_rows + promoted_rows
    with open(PUBLIC_FOUNDING_PATH, 'w', encoding='utf-8') as f:
        f.write(','.join(OUTPUT_HEADERS) + '\n')
        for row in output_rows:
            f.write(','.join(csv_escape(row.get(h, '')) for h in OUTPUT_HEADERS) + '\n')

    print(json.dumps({'promoted': len(promoted_rows), 'total_live_rows': len(output_rows)}, indent=2))


if __name__ == '__main__':
    main()
